src/app/reserve_queue/reserve_queue.cpp

```diff
@@ -0,0 +1,95 @@
+#include "reserve_queue.hpp"
+#include <algorithm>
+
+namespace
+{
+bool addPatientIfNotPresent(ReserveQueue::PatientQueueType& patients_queued, const std::shared_ptr<Patient>& patient)
+{
+    auto patient_iterator = std::ranges::find(patients_queued, patient);
+
+    if (patient_iterator == patients_queued.end())
+    {
+        patients_queued.emplace_back(patient);
+
+        return true;
+    }
+
+    return false;
+}
+
+void removePatientFromQueue(ReserveQueue::PatientQueueType& patients_queued, const std::shared_ptr<Patient>& patient)
+{
+    if (auto patientSearched = std::ranges::find(patients_queued, patient); patientSearched != patients_queued.end())
+    {
+        patients_queued.erase(patientSearched);
+    }
+}
+} // namespace
+
+bool ReserveQueue::addAwaitingPatient(Treatment treatment, const std::shared_ptr<Patient>& patient)
+{
+    return addPatientIfNotPresent(treatment_queue_[treatment], patient);
+}
+
+bool ReserveQueue::addAwaitingPatient(const std::shared_ptr<Doctor>& doctor, const std::shared_ptr<Patient>& patient)
+{
+    return addPatientIfNotPresent(doctor_queue_[doctor], patient);
+}
+
+void ReserveQueue::removeAwaitingPatient(Treatment treatment, const std::shared_ptr<Patient>& patient)
+{
+    auto& patients_queued = treatment_queue_[treatment];
+
+    removePatientFromQueue(patients_queued, patient);
+
+    if (patients_queued.empty())
+    {
+        treatment_queue_.erase(treatment);
+    }
+}
+
+void ReserveQueue::removeAwaitingPatient(const std::shared_ptr<Doctor>& doctor, const std::shared_ptr<Patient>& patient)
+{
+    auto& patients_queued = doctor_queue_[doctor];
+
+    removePatientFromQueue(patients_queued, patient);
+
+    if (patients_queued.empty())
+    {
+        doctor_queue_.erase(doctor);
+    }
+}
+
+ReserveQueue::PatientQueueType ReserveQueue::awaitingPatients(Treatment treatment) const
+{
+    auto queued_patients = treatment_queue_.find(treatment);
+
+    return queued_patients != treatment_queue_.end() ? queued_patients->second : PatientQueueType{};
+}
+
+ReserveQueue::PatientQueueType ReserveQueue::awaitingPatients(const std::shared_ptr<Doctor>& doctor) const
+{
+    auto queued_patients = doctor_queue_.find(doctor);
+
+    return queued_patients != doctor_queue_.end() ? queued_patients->second : PatientQueueType{};
+}
+
+std::vector<Treatment> ReserveQueue::treatmentsHavingQueue() const
```

Comment:
`getTreatmentsQueue `

Reply:
IMHO, I do not agree with this name proposal.
Here we return just a vector of enums being keys for possible queues extraction - not the actual queues.
proposed name ("getTreatmentsQueue") might mislead someone that actual queue might be received.

I realize that name is not perfect :)
I spent at least 20 minutes trying to figure out names for these and "doctorsHavingQueue" function name.

I could not figure out anything better, which could explain function best :(
I am open for proposals anyway, although I do not like this particular one ;)